[markusyeo] ip

markusyeo opened the pull request:
**Not Duke, but Joe**
> “You shouldn't work so hard, you should sleep early.” – Prof Wai Kay(CS1010S)

Remember this is not Duke. but Joe! Joe is your personal butler in your pocket who can remember what you have to do.

- text-based
- somewhat easy to learn
- just regularly FAST to use

All you need to do is,

download it from somewhere on my git page
2. add your tasks.
3. let it manage your tasks for you 😉
And it is **FREE** (probably the only good thing about it)!

Features:

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:

```
public static void main(String[] args) {
        ...
        String logo = "     |             \n"
                           + "     |  _ \\    _ \\ \n"
                          + " \\   | (   |   __/ \n"
                           + "\\___/ \\___/  \\___|\n";
```

Commit:
Add Gradle support

markusyeo committed:
Level-1

markusyeo committed:
Level-2

markusyeo committed:
Level-3

markusyeo committed:
Level-4

markusyeo committed:
Level-4

markusyeo committed:
A-TextUiTesting

markusyeo committed:
Level-5

markusyeo committed:
Level-6

markusyeo committed:
Update data structure of Duke

markusyeo committed:
Add Level-7 features

markusyeo committed:
Add Level-8 datetime feature

markusyeo committed:
Merge branch 'branch-level-7'

markusyeo committed:
Merge branch 'branch-level-8'

markusyeo committed:
Add package

markusyeo committed:
Merge remote-tracking branch 'origin/add-gradle-support'

markusyeo committed:
Update build.gradle file

markusyeo committed:
Add checkstyle support

markusyeo committed:
Add find

markusyeo committed:
Add Junit testing
- Fix Parser bugs
- Fix Event toString method

markusyeo committed:
Make checkstyle happy

markusyeo committed:
Add javadoc

markusyeo committed:
Merge remote-tracking branch 'origin/branch-Level-9'

markusyeo committed:
Merge remote-tracking branch 'origin/branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Event.java

markusyeo committed:
Merge remote-tracking branch 'origin/branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Deadline.java
#	src/main/java/duke/Task.java
#	src/main/java/duke/ToDo.java

markusyeo committed:
Fix Event.java

Yaladah commented on src/main/java/duke/Duke.java:
```diff
@@ -1,8 +1,17 @@
+     *
+     * @param args Input arguments, will be ignored.
+     */
     public static void main(String[] args) {
```
Maybe you could consider abstracting out the behaviour of the bot into a method run()?

gremmyz replied:
Think I agree with this, I believe some lines here can be abstracted out into the Joe class.

Yaladah reviewed (commented):
I like that I could find basically no coding violations in the code. It might be nice to abstract out a Ui class that handles interaction with the user, such as printing of messages and taking in inputs from the user. (It seems to be a requirement for the ip too if I'm not mistaken.)

gremmyz commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,31 @@
+ */
+public class Event extends Task {
+
+    private final LocalDateTime fromTime;
+    private final LocalDateTime toTime;
```
Good job with the descriptive naming!

gremmyz commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,41 @@
+        super(msg);
+    }
+
+    public static void rethrow(String s) throws ToDoException, UnknownCommandException {
+        try {
+            if (s.equals("ToDoException")) {
+                throw new ToDoException();
+            } else {
+                throw new UnknownCommandException();
+            }
+        } catch (Exception e) {
+            throw e;
+        }
+    }
```
Really cool way of handling exceptions 👍🏼

gremmyz commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,136 @@
+        this.taskList = taskList;
+    }
+    String parse(String input) {
+        if (input.equals("bye")) {
+            return sayBye();
+        } else if (input.equals("list")) {
+            return returnList(taskList);
+        } else if (input.startsWith("mark")) {
+            return mark(input);
+        } else if (input.startsWith("unmark")) {
+            return unmark(input);
+        } else if (input.startsWith("delete")) {
+            return delete(input);
+        } else if (input.startsWith("find")) {
+            return find(input);
+        } else {
+            return addTask(input)[1];
+        }
```
Out of curiousity, any reason why you chose to use _if-else_ here instead of _switch_ like below?

gremmyz commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,136 @@
+                return new String[]{String.valueOf(taskList.size()), String.format("added: %s", task)};
+            }
+            case "d": {
+                DateTimeFormatter altFormatter = DateTimeFormatter.ofPattern("EEE, d MMM yyyy hh:mma");
+                String[] s = split[1].split(" by: ");
+                System.out.println(s[0]);
+                LocalDateTime byTime = LocalDateTime.parse(s[1], altFormatter);
+                Task task = new Deadline(s[0], byTime);
+                taskList.add(task);
+                return new String[]{String.valueOf(taskList.size()), String.format("added: %s", task)};
+            }
+            case "deadline": {
+                String[] s = split[1].split(" /by ");
+                System.out.println(s[0]);
+                LocalDateTime byTime = LocalDateTime.parse(s[1], formatter);
+                Task task = new Deadline(s[0], byTime);
+                taskList.add(task);
+                return new String[]{String.valueOf(taskList.size()), String.format("added: %s", task)};
+            }
+            case "e": {
+                DateTimeFormatter altFormatter = DateTimeFormatter.ofPattern("EEE, d MMM yyyy hh:mma");
+                String[] s = split[1].split(" from: ", 2);
+                String taskDescription = s[0];
+                s = s[1].split(" to: ");
+                LocalDateTime fromTime = LocalDateTime.parse(s[0], altFormatter);
+                LocalDateTime toTime = LocalDateTime.parse(s[1], altFormatter);
+                Task task = new Event(taskDescription, fromTime, toTime);
+                taskList.add(task);
+                return new String[]{String.valueOf(taskList.size()), String.format("added: %s", task)};
+            }
+            case "event": {
+                String[] s = split[1].split(" /from ", 2);
+                String taskDescription = s[0];
+                s = s[1].split(" /to ");
+                LocalDateTime fromTime = LocalDateTime.parse(s[0], formatter);
+                LocalDateTime toTime = LocalDateTime.parse(s[1], formatter);
+                Task task = new Event(taskDescription, fromTime, toTime);
+                taskList.add(task);
+                return new String[]{String.valueOf(taskList.size()), String.format("added: %s", task)};
+            }
```
Any reason why you use one method to handle addTask inputs from both the CLI and the saved file?

gremmyz commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,136 @@
+        return out.toString();
+    }
+
+    String mark(String input) {
+        int i = Integer.parseInt(input.split(" ", 2)[1]) - 1;
+        Task task = taskList.get(i);
+        task.mark();
+        return String.format("\tNice! I've marked this task as done:\n\t  %s", task);
+    }
+
+    private String unmark(String input) {
+        int i = Integer.parseInt(input.split(" ", 2)[1]) - 1;
+        Task task = taskList.remove(i);
+        task.unmark();
+        return String.format("\tNice! I've marked this task as not done yet:\n\t  %s", task);
+    }
```
Not sure if it's a missing access modifier for `mark` so I thought I'd let you know in case.

gremmyz reviewed (commented):
LGTM! Overall, I think your code is really clean, good job! Just missing a few javadocs here and there, may want to look into it if you have time! 😄

markusyeo committed:
Add Ui

markusyeo committed:
Add GUI
- fix storage

markusyeo committed:
Add assertions

The Storage class handles the reading, writing and parsing of the storage text file and handles corrupt storage with exception handling.

The Storage class can use some assert statement handling.

Let's add some assert statements where possible.

This will improve the QOL of the code and it will be happier.

markusyeo committed:
Merge pull request #2 from markusyeo/A-Assertions

Add assertions

markusyeo committed:
Improve code quality

The Parser class code is ridiculously long, and a
little messy. It is difficult to properly read through the addTask method.

Splitting up the switch cases into their own methods
will neaten things up and improve readability.

Let's,
* extract addToDo, addDeadline and addEvent
* extract the DateTimeFormatter to a private variable

This refactoring will make the code more navigable
and make it easier to add more features in the
future.

markusyeo committed:
Improve more code quality

markusyeo committed:
Merge pull request #3 from markusyeo/A-CodeQuality

A code quality update

markusyeo committed:
Add bcd-extension - sort

markusyeo committed:
Fix Storage and Parser classes

markusyeo committed:
Remove unnecessary variable definition in Storage class

markusyeo committed:
Fix sort functionality

markusyeo committed:
Change mainClassName in build.gradle

markusyeo committed:
Fix Joe
- Update some text formatting
- Add Ui.png
- Add Userguide

markusyeo committed:
Checkstyle and javadocs

markusyeo committed:
Add Circle CI
